source/clustering.md

```diff
@@ -374,37 +444,121 @@ we sum them together to get the *total WSSD*.
 to each of the K clusters.
-An example random initialization is shown in {numref}`toy-kmeans-init-1`.
+An example random initialization is shown in {numref}`toy-kmeans-init-1`
 
 
+:::{glue:figure} toy-kmeans-init-1
+:figwidth: 700px
+:name: toy-kmeans-init-1
 
-```{figure} img/clustering/toy-kmeans-init-1.png
----
-height: 400px
-name: toy-kmeans-init-1
----
 Random initialization of labels.
+:::
+
+```{code-cell} ipython3
+:tags: [remove-cell]
+
+from sklearn.cluster import KMeans
 
+
+def plot_kmean_iterations(iterations, data, centroid_init):
+    """Plot kmeans cluster and label updates for multiple iterations"""
+    dfs = []
+    centroid_inits = []
+    for i in range(1, iterations+1):
+        data['iteration'] = f'Iteration {i}'
+        data['update_type'] = 'Center Update'
+        data['flipper_centroid'] = data['label'].map(centroid_init['flipper_length_standardized'])
+        data['bill_centroid'] = data['label'].map(centroid_init['bill_length_standardized'])
+        dfs.append(data.copy())
+
+        cluster_columns = ['flipper_length_standardized', 'bill_length_standardized']
+        knn = KMeans(init=centroid_init[cluster_columns], n_clusters=3, max_iter=1, n_init=1)
+        knn.fit(data[cluster_columns])
+
+        data['label'] = knn.labels_
+        data['iteration'] = f'Iteration {i}'
+        data['update_type'] = 'Label Update'
+        data['flipper_centroid'] = data['label'].map(centroid_init['flipper_length_standardized'])
+        data['bill_centroid'] = data['label'].map(centroid_init['bill_length_standardized'])
+        dfs.append(data.copy())
+
+        centroid_init = data.groupby('label').mean(numeric_only=True)
+
+    points = alt.Chart(
+        pd.concat(dfs),
+        width=200,
+        height=200
+    ).mark_point(filled=True, size=75, opacity=1).encode(
+        alt.X("flipper_length_standardized").scale(domain=(-2, 2)),
+        alt.Y("bill_length_standardized").scale(domain=(-2, 2)),
+        alt.Color('label:N').legend(None),
+        alt.Shape('label:N').legend(None)
+    )
+
+    centroids = points.mark_point(size=200, filled=True, stroke='black', strokeWidth=1.5).encode(
+        alt.X("mean(flipper_centroid)")
+            .scale(domain=(-2, 2))
+            .title("Flipper Length (standardized)"),
+        alt.Y("mean(bill_centroid)")
+            .scale(domain=(-2, 2))
+            .title("Flipper Length (standardized)")
+    )
+
+    return (points + centroids).facet(
+        row=alt.Row('iteration', header=alt.Header(title='', labelFontSize=18)),
+        column=alt.Column('update_type', header=alt.Header(title='', labelFontSize=18))
+    )
+```
+
+```{code-cell} ipython3
+:tags: [remove-cell]
+
+centroid_init = penguin_data.groupby('label').mean()
+# For some reason, this adjustment is needed to get roughly the same assignment of the points
+# to the clusters as in the R version
+centroid_init.loc[2] = [0.29, 0.36]
+centroid_init.loc[1] = [-0.01, -0.09]
```

Comment:
I am not sure why this is the case, but I didn't look too closely at the R code (although it seemed like they used the mean there). Not that there is still one point that is different but it has not impact on the take home message